### PR TITLE
Runtime behavior for entitlements

### PR DESCRIPTION
**Pull request by dsainati1**
Part of the implementation of https://github.com/onflow/flips/pull/54, part of https://github.com/onflow/cadence/issues/2412

Implements the runtime behavior for entitlements at runtime

______

- [ ] Targeted PR against `master` branch
- [X] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation 
- [X] Re-reviewed `Files changed` in the Github PR explorer
- [X] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter.go [hunk 1]
```diff
@@ -4431,6 +4508,62 @@ func (interpreter *Interpreter) ReportComputation(compKind common.ComputationKin
...
+
+func (interpreter *Interpreter) mapMemberValueAuthorization(self Value, memberAccess *sema.Access, resultValue Value) Value {
+	if memberAccess == nil {
+		return resultValue
+	}
+	if mappedAccess, isMappedAccess := (*memberAccess).(sema.EntitlementMapAccess); isMappedAccess {
+		var auth EntitlementSetAuthorization
+		switch selfValue := self.(type) {
+		case AuthorizedValue:
+			imageAccess, err := mappedAccess.Image(interpreter.MustConvertStaticAuthorizationToSemaAccess(selfValue.GetAuthorization()), ast.EmptyRange)
```
Comment (SupunS) [inline runtime/interpreter/interpreter.go:4535]: could break this into two lines
```suggestion
			selfAccess := interpreter.MustConvertStaticAuthorizationToSemaAccess(selfValue.GetAuthorization())
			imageAccess, err := mappedAccess.Image(selfAccess, ast.EmptyRange)
```

**Review by SupunS — COMMENTED** [review]
Reviewed up to tests. Will check the remaining in a second batch.

File: runtime/tests/checker/entitlements_test.go [hunk 2]
```diff
@@ -128,9 +129,10 @@ func TestCheckBasicEntitlementMappingNonEntitlements(t *testing.T) {
 			}
 		`)
 
-		errs := RequireCheckerErrors(t, err, 1)
+		errs := RequireCheckerErrors(t, err, 2)
 
-		require.IsType(t, &sema.InvalidNonEntitlementTypeInMapError{}, errs[0])
+		require.IsType(t, &sema.NotDeclaredError{}, errs[0])
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/tests/checker/entitlements_test.go:134]: Basically because the entitlement mappings have to be processed by the checker before struct definitions (so they can be used in struct definitions), meaning that if someone puts a struct type inside an entitlement mapping (which doesn't work anyways) the checker doesn't know about that struct yet.

File: runtime/interpreter/statictype.go [hunk 3]
```diff
@@ -511,41 +512,47 @@ func NewEntitlementSetAuthorization(memoryGauge common.MemoryGauge, entitlements
...
+			if !slices.Contains(auth.Entitlements, entitlement) {
 				return false
 			}
 		}
-		return e.Kind == auth.Kind
+		for _, entitlement := range auth.Entitlements {
+			if !slices.Contains(e.Entitlements, entitlement) {
+				return false
+			}
+		}
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/interpreter/statictype.go:554]: In the general case that's probably true, but I expect that these entitlement sets will be very small in practice (2-3 elements usually) that the overhead of a map is unlikely to be worth it.

File: runtime/interpreter/value.go [hunk 4]
```diff
@@ -14962,12 +14963,6 @@ func (v *CompositeValue) Transfer(
 				// and does not need to be converted or copied
 
 				value := MustConvertStoredValue(interpreter, atreeValue)
-				// the base of an attachment is not stored in the atree, so in order to make the
-				// transfer happen properly, we set the base value here if this field is an attachment
-				if compositeValue, ok := value.(*CompositeValue); ok && compositeValue.Kind == common.CompositeKindAttachment {
-					compositeValue.setBaseValue(interpreter, v)
-				}
-
```
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/interpreter/value.go:14970]: I don't think it ever was tbh, I realized during some debugging that the base value was always already set by the time this check was reached, so I just removed it.

File: runtime/interpreter/statictype.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dsainati1) [reply runtime/interpreter/statictype.go]: Discussed this in a meeting, will replace this slice with an ordered set.

**Review by SupunS — APPROVED** [review]

File: encoding/ccf/encode_type.go [hunk 5]
```diff
@@ -318,6 +318,13 @@ func (e *Encoder) encodeReferenceType(
 	)
 }
 
+func (e *Encoder) encodeAuthorization(
+	auth cadence.Authorization,
+) error {
+	// TODO: implement this
```
Comment (turbolent) [inline encoding/ccf/encode_type.go:324]: Is this implemented in a subsequent PR?

File: runtime/interpreter/statictype.go [hunk 6]
```diff
@@ -605,8 +613,6 @@ func (e EntitlementMapAuthorization) Equal(auth Authorization) bool {
 // ReferenceStaticType
 
 type ReferenceStaticType struct {
-	// BorrowedType is the type of the usage (T in &T)
-	BorrowedType StaticType
 	// ReferencedType is type of the referenced value (the type of the target)
 	ReferencedType StaticType
```
Comment (turbolent) [inline runtime/interpreter/statictype.go:617]: Why was the borrowed type removed? Isn't it still needed?

File: runtime/sema/access.go [hunk 7]
```diff
@@ -296,7 +296,7 @@ func (e EntitlementMapAccess) Domain() EntitlementSetAccess {
 	for _, relation := range e.Type.Relations {
 		domain[relation.Input] = struct{}{}
 	}
-	e.domain = NewEntitlementSetAccess(maps.Keys(domain), Disjunction)
+	e.domain = NewEntitlementSetAccess(maps.Keys(domain), Conjunction)
```
Comment (turbolent) [inline runtime/sema/access.go:299]: Why did this change?

File: runtime/storage_test.go [hunk 8]
```diff
@@ -1556,9 +1556,7 @@ func TestRuntimeStorageReferenceCast(t *testing.T) {
 		},
 	)
 
-	RequireError(t, err)
-
-	require.ErrorAs(t, err, &interpreter.ForceCastTypeMismatchError{})
+	require.NoError(t, err)
```
Comment (turbolent) [inline runtime/storage_test.go:1559]: Maybe add a test case that ensures that the type cannot be downcast with additional entitlements

File: runtime/interpreter/encoding_test.go [hunk 9]
```diff
@@ -3536,9 +3531,9 @@ func TestEncodeDecodePathLinkValue(t *testing.T) {
 			// authorization
 			// tag
 			0xd8, CBORTagEntitlementSetStaticAuthorization,
-			// array, 2 items follow
+			// array, length 2
 			0x82,
-			// Disjunction
+			// conjunction
```
Comment (turbolent) [inline runtime/interpreter/encoding_test.go:3536]: Accident?

File: runtime/tests/interpreter/runtimetype_test.go [hunk 10]
```diff
@@ -505,24 +505,22 @@ func TestInterpretReferenceType(t *testing.T) {
       struct S {}
 	  entitlement X
 
-      let a = ReferenceType(authorized: true, type: Type<@R>())
-      let b = ReferenceType(authorized: false, type: Type<String>())
-      let c = ReferenceType(authorized: true, type: Type<S>())
+      let a = ReferenceType(entitlements: ["S.test.X"], type: Type<@R>())!
```
Comment (turbolent) [inline runtime/tests/interpreter/runtimetype_test.go:508]: Should the entitlements set maybe be `[Type]` instead of `[String]`?

File: encoding/ccf/encode_type.go (hunk 5 above)
Comment (dsainati1) [reply encoding/ccf/encode_type.go:324]: No, I spoke to @fxamacker about this and added this TODO (and created an issue) for her to track it; she wants to implement this herself in the future.

Comment (dsainati1) [reply encoding/ccf/encode_type.go:324]: No, I spoke to @fxamacker about this and added this TODO (and created an issue) for her to track it; she wants to implement this herself in the future.

File: runtime/tests/interpreter/runtimetype_test.go (hunk 10 above)
Comment (dsainati1) [reply runtime/tests/interpreter/runtimetype_test.go:508]: No, there is no runtime function to create a runtime entitlement type, and IMO there should not be. They aren't really types, and it would be confusing for you to be able to create a `Type` value for an entitlement (especially because they are "erased" at runtime).

File: runtime/interpreter/statictype.go (hunk 6 above)
Comment (dsainati1) [reply runtime/interpreter/statictype.go:617]: No, the "BorrowedType" was previously used to check what the "visible" type of the reference was, for things like casting where the type behavior of the reference was not necessarily the same as the type behavior of its underlying type. This has been simplified, a reference's borrow type is the same as the type of its underlying value at runtime (i.e. downcasting is possible), so we don't need this anymore.

Comment (turbolent) [reply runtime/interpreter/statictype.go:617]: I see, that makes sense. My confusion probably stemmed from the fact that I assumed we would keep `BorrowedType`, because that was the actual "static type", and remove `ReferencedType` as that was the "dynamic type". 

It would be good to keep the comment though: `ReferencedType is the type T in &T`. The current comment, "referenced value (the type of the target)" meant that it was actually the concrete type of the value that the reference value was pointing to. That is not the case anymore (I assume?)

File: runtime/sema/access.go (hunk 7 above)
Comment (turbolent) [reply runtime/sema/access.go]: Sounds good, just wanted to make sure it was not an accident

File: runtime/interpreter/value.go [hunk 11]
```diff
@@ -17374,6 +17435,15 @@ func (v *StorageReferenceValue) GetTypeKey(
 ) Value {
 	self := v.mustReferencedValue(interpreter, locationRange)
 
+	if selfComposite, isComposite := self.(*CompositeValue); isComposite {
```
Comment (turbolent) [inline runtime/interpreter/value.go:17438]: Why is this special case for `*CompositeValue` needed?

File: runtime/interpreter/statictype.go (hunk 6 above)
Comment (turbolent) [reply runtime/interpreter/statictype.go:617]: Could we remove `ReferencedType` instead of `BorrowedType`? I have a bad feeling we currently use `BorrowedType` somewhere, initialize `ReferencedType` to a different value, and then the PR switches the use to `ReferencedType` and suddenly the semantics are wrong. It would reduce the number of changes in the PR

File: runtime/interpreter/interpreter_expression.go [hunk 12]
```diff
@@ -1076,14 +1077,12 @@ func (interpreter *Interpreter) VisitCastingExpression(expression *ast.CastingEx
 			}
 
 			// The failable cast may upcast to an optional type, e.g. `1 as? Int?`, so box
-			value = interpreter.BoxOptional(locationRange, value, expectedType)
+			value = interpreter.ConvertAndBox(locationRange, value, valueSemaType, expectedType)
```
Comment (turbolent) [inline runtime/interpreter/interpreter_expression.go:1080]: I guess this change is unrelated to entitlements? (vaguely remember we discussed this before?)

File: runtime/interpreter/interpreter_expression.go [hunk 13]
```diff
@@ -1140,6 +1139,25 @@ func (interpreter *Interpreter) VisitReferenceExpression(referenceExpression *as
 
 	interpreter.maybeTrackReferencedResourceKindedValue(result)
 
+	makeReference := func(value Value, typ *sema.ReferenceType) *EphemeralReferenceValue {
```
Comment (turbolent) [inline runtime/interpreter/interpreter_expression.go:1142]: Nice refactor!

File: runtime/sema/check_function.go [hunk 14]
```diff
@@ -393,7 +393,7 @@ func (checker *Checker) visitWithPostConditions(postConditions *ast.Conditions,
 			// here the `result` value in the `post` block will have type `auth(E, X, Y) &R`
 			if entitlementSupportingType, ok := innerType.(EntitlementSupportingType); ok {
 				supportedEntitlements := entitlementSupportingType.SupportedEntitlements()
-				if supportedEntitlements == nil || supportedEntitlements.Len() > 0 {
+				if supportedEntitlements != nil && supportedEntitlements.Len() > 0 {
```
Comment (turbolent) [inline runtime/sema/check_function.go:396]: Why did this change? Do we have a test case?

**Review by turbolent — COMMENTED** [review]
Great work! Pretty close now 👏

File: runtime/interpreter/value.go (hunk 11 above)
Comment (dsainati1) [reply runtime/interpreter/value.go:17438]: The implementation of `GetTypeKey` for `CompositeValue`s is written under the assumption that it is called with an owned value, and calls into `getTypeKey` with the full domain of the attachment's entitlement map as the attachment's access. This is not the behavior we want when we are accessing an attachment on a reference to a composite, so we compute the entitlement set of the reference and call `getTypeKey` with that access instead.

File: runtime/sema/check_function.go (hunk 14 above)
Comment (dsainati1) [reply runtime/sema/check_function.go:396]: This I think is left over from a previous PR; `SupportedEntitlements` used to return `nil` and now it cannot, so these two cases are actually equivalent. I just removed the first condition to simplify.

File: runtime/interpreter/interpreter_expression.go (hunk 12 above)
Comment (dsainati1) [reply runtime/interpreter/interpreter_expression.go:1080]: It is related yes; the reason for this change is that we want casting an entitled reference to always "convert" the set of entitlements on the runtime value to the target type of the cast; i.e. it shouldn't be possible to "hide" an entitlement at runtime. It's the difference between this code having `downDownRef` be `nil` and non-`nil`:

```
let ref = &1 as auth(X) &Int
let anyStruct = ref as AnyStruct
let downRef = (anyStruct as? &Int)!
let downDownRef = downRef as? auth(X) &Int
```

File: runtime/interpreter/statictype.go (hunk 6 above)
Comment (dsainati1) [reply runtime/interpreter/statictype.go:617]: The comment as written is accurate; `ReferencedType` is the type of the referenced value. The type of the `T` that appears in the `&T` is no longer part of the static type, as it is not necessary, as we can now make all up and downcasting decisions based on the runtime type of the referenced value, and do not need to consider the type it was originally borrowed with. The exception to this are the entitlements of the reference, which are tracked separately. No other static types keep a converted copy of the original sema type around; references only did this because previously the runtime value was only allowed to be used in ways that reflected its original borrow type. 

The only places in the code where this actually comes up are in the interpreter encoding (which used to encode the `BorrowedType` and now uses the `ReferencedType`), in `IsSubtypeOfSemaType` in the interpreter (which was changed before this field was removed, and was the impetus for removing it in the first place), and in the creation of sema types from static types for references (which is also intentional).

Comment (turbolent) [reply runtime/interpreter/statictype.go:617]: Thank you for the great explanation! 👍

**Review by turbolent — APPROVED** [review]
👏 🚀
